src/Bicep.Core.IntegrationTests/UserDefinedTypeTests.cs

```diff
@@ -400,23 +402,26 @@ public void Additional_properties_may_be_used_alongside_named_properties()
     [TestMethod]
     public void Nullably_typed_values_can_be_used_as_nonnullable_outputs_with_postfix_assertion()
     {
-        var result = CompilationHelper.Compile(ServicesWithUserDefinedTypes, @"
+        var templateWithPossiblyNullDeref = @"
 param foos (null | { bar: { baz: { quux: 'quux' } } })[]
 
 output quux string = foos[0].bar.baz.quux
-");
+";
+        var templateWithNonNullAssertion = @"
+param foos (null | { bar: { baz: { quux: 'quux' } } })[]
 
+output quux string = foos[0]!.bar.baz.quux
+";
+
+        var result = CompilationHelper.Compile(ServicesWithUserDefinedTypes, templateWithPossiblyNullDeref);
         result.Should().HaveDiagnostics(new []
         {
-          ("BCP055", DiagnosticLevel.Error, @"Cannot access properties of type ""null | { bar: { baz: { quux: 'quux' } } }"". An ""object"" type is required."),
+          ("BCP318", DiagnosticLevel.Warning, @"The value of type ""null | { bar: { baz: { quux: 'quux' } } }"" may be null at deploy time, which would cause the deployment to fail."),
```

Comment:
I remember @anthony-c-martin once suggested that we avoid using terms like "deploy time" and "runtime" in error messages since not all users understand the concepts. That's why the DTC validation error messages (e.g. BCP120) say "at the start of the deployment".  Do we want to use the same phrase here?

Reply:
Updated in https://github.com/Azure/bicep/pull/9596/commits/ceb61780059c16b8660cc841e6a76cfaf10529d0